File: src/services/audit/detectors/ad/computers/other/pre-windows-2000.ts
/**
 * Computer Pre-Windows 2000 Detector
 * Check for Pre-Windows 2000 computer accounts
 */

import { ADComputer } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedComputerEntities, ldapAttrToString } from '../../../../../../utils/entity-converter';

export function detectComputerPreWindows2000(computers: ADComputer[], includeDetails: boolean): Finding {
  const affected = computers.filter((c) => {
    const os = ldapAttrToString(c.operatingSystem);
    if (!os) return false;
    return /Windows NT|Windows 2000|Windows 95|Windows 98/i.test(os);
  });

  return {
    type: 'COMPUTER_PRE_WINDOWS_2000',
    severity: 'medium',
    category: 'computers',
    title: 'Pre-Windows 2000 Computer',
    description: 'Pre-Windows 2000 compatible computer. Weak security settings, potential compatibility exploits.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedComputerEntities(affected) : undefined,
  };
}

File: internal/audit/detectors/ad/computers/other/pre-windows-2000.go
package other

import (
	"context"
	"regexp"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// PreWindows2000Detector checks for pre-Windows 2000 computers
type PreWindows2000Detector struct {
	audit.BaseDetector
}

// NewPreWindows2000Detector creates a new detector
func NewPreWindows2000Detector() *PreWindows2000Detector {
	return &PreWindows2000Detector{
		BaseDetector: audit.NewBaseDetector("COMPUTER_PRE_WINDOWS_2000", audit.CategoryComputers),
	}
}

var preWin2000Pattern = regexp.MustCompile(`(?i)Windows NT|Windows 2000|Windows 95|Windows 98`)

// Detect executes the detection
func (d *PreWindows2000Detector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.Computer

	for _, c := range data.Computers {
		if c.OperatingSystem == "" {
			continue
		}
		if preWin2000Pattern.MatchString(c.OperatingSystem) {
			affected = append(affected, c)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Pre-Windows 2000 Computer",
		Description: "Pre-Windows 2000 compatible computer. Weak security settings, potential compatibility exploits.",
		Count:       len(affected),
	}

	if len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedComputerEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewPreWindows2000Detector())
}
